src/words/environment/WordsObject.java

```diff
@@ -13,13 +15,35 @@
 	private Position cell;
 	private String currentMessage;
 	private Action lastAction;
-
+	private HashMap<WordsObject, ArrayList<Property>> referers;
+	
+	public void updateReferers(WordsObject referringObject, Property referringProperty) {
+		ArrayList<Property> referringProperties;
+		if ((referringProperties = referers.get(referringObject)) == null) {
+			referringProperties = new ArrayList<Property>();
+			referers.put(referringObject, referringProperties);
+		}
+		
+		referringProperties.add(referringProperty);
+	}
+	
+	public void clearReferers() {
+		for (WordsObject referringObject : referers.keySet()) {
+			for (Property property : referers.get(referringObject)) {
+				property.type = PropertyType.NOTHING;
+				property.objProperty = null;
+			}
+			referers.remove(referringObject);
```

Comment:
Since we're modifying the list of referers while we're in the loop, I think we need to use the iterator syntax rather than this "enhanced for loop" shorthand.  You can see how Wangda did it in event listeners in the frame loop.  See this post for details: http://stackoverflow.com/questions/1196586/calling-remove-in-foreach-loop-in-java.